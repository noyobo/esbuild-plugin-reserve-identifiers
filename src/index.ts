import fs from 'fs';
import path from 'path';
import { Loader, Plugin } from 'esbuild';

const loaderMap: Record<string, Loader> = {
  '.ts': 'ts',
  '.tsx': 'tsx',
  '.js': 'js',
  '.jsx': 'jsx',
  '.mjs': 'js',
};

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isString(value: unknown): value is string {
  return typeof value === 'string';
}

function isArray(value: unknown): value is unknown[] {
  return Array.isArray(value);
}

function slash(path: string) {
  const isExtendedLengthPath = path.startsWith('\\\\?\\');

  if (isExtendedLengthPath) {
    return path;
  }

  return path.replace(/\\/g, '/');
}

export function reserveIdentifiersPlugin(
  options:
    | { filter?: RegExp; identifiers: string[] }
    | {
        filter?: RegExp;
        /**
         * @deprecated Use `identifiers` instead
         */
        reserveIdentifiers: string[];
      }
) {
  return {
    name: 'reserve-identifiers',
    setup(build) {
      const identifiers = 'reserveIdentifiers' in options ? options.reserveIdentifiers : options.identifiers;
      const reserveIdentifiersCode = `(${identifiers.map((key) => `typeof ${key}`).join(',')});\n`;

      const initialOptions = build.initialOptions;
      const entryPoints = initialOptions.entryPoints;

      function isEntryPoint(filePath: string) {
        if (isObject(entryPoints)) {
          return Object.values(entryPoints).some((entryPoint) => {
            if (isString(entryPoint)) {
              return slash(path.resolve(entryPoint)) === slash(filePath);
            }
            return false;
          });
        }
        if (isArray(entryPoints)) {
          return entryPoints.some((entryPoint) => {
            if (isString(entryPoint)) {
              return slash(path.resolve(entryPoint)) === slash(filePath);
            }
            if (isObject(entryPoint)) {
              const inputPath = entryPoint.in;
              return slash(path.resolve(inputPath)) === slash(filePath);
            }
            return false;
          });
        }
      }

      build.onLoad({ filter: options?.filter || /\.([tj]sx?|mjs)$/ }, async (args) => {
        if (isEntryPoint(args.path)) {
          const extname = path.extname(args.path);
          const content = fs.readFileSync(args.path, 'utf-8');

          const code = reserveIdentifiersCode + content;
          return { contents: code, watchFiles: [args.path], loader: loaderMap[extname] || 'js' };
        }
        return null;
      });
    },
  } as Plugin;
}
