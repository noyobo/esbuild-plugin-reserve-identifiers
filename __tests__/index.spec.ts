import * as path from 'path';
import esbuild, { BuildResult } from 'esbuild';
import { reserveIdentifiersPlugin } from '../src/index';

const createBuild = async (files: string[]) => {
  files = files.map((file) => {
    return path.join(__dirname, file);
  });

  await esbuild
    .build({
      entryPoints: files,
      plugins: [
        reserveIdentifiersPlugin({
          identifiers: [
            'ty',
            'wx',
            'a',
            'b',
            'c',
            'd',
            'e',
            'f',
            'g',
            'h',
            'i',
            'j',
            'k',
            'l',
            'm',
            'n',
            'o',
            'p',
            'q',
            'r',
            's',
            't',
            'u',
            'v',
            'w',
            'x',
            'y',
            'z',
            'A',
            'B',
            'C',
            'D',
            'E',
            'F',
            'G',
            'H',
            'I',
            'J',
            'K',
            'L',
            'M',
            'N',
            'O',
            'P',
            'Q',
            'R',
            'S',
            'T',
            'U',
            'V',
            'W',
            'X',
            'Y',
            'Z',
            '_',
            '$',
          ],
        }),
      ],
      write: false,
      bundle: true,
      target: 'es6',
      sourcemap: false,
      outbase: __dirname,
      outdir: 'dist',
      minify: true,
      external: ['react', 'react-dom'],
    })
    .then((result) => {
      expect(result.errors.length).toBe(0);
      expect(result.outputFiles.map((f) => f.text)).toMatchSnapshot();
    })
    .catch((error: BuildResult) => {
      expect(error.errors.length).toBe(1);
      expect(error.errors[0].text).toContain('Expected ")" but found ";"');
    });
};

describe('index', () => {
  describe('esbuild-plugin-reserve-identifiers', () => {
    it('basic', async () => {
      await createBuild(['./fixtures/basic/index.ts']);
    });

    it('react', async () => {
      await createBuild(['./fixtures/react/index.tsx']);
    });

    it('react-jsx', async () => {
      await createBuild(['./fixtures/react-jsx/index.jsx']);
    });

    it('async', async () => {
      await createBuild(['./fixtures/async/index.ts']);
    });

    it('error', async () => {
      await createBuild(['./fixtures/error/index.ts']);
    });
  });
});
