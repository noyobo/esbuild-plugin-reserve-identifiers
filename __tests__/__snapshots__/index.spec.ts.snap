// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index esbuild-plugin-reserve-identifiers async 1`] = `
[
  ""use strict";(()=>{var eo=(fo,po,to)=>new Promise((co,no)=>{var mo=oo=>{try{yo(to.next(oo))}catch(ro){no(ro)}},bo=oo=>{try{yo(to.throw(oo))}catch(ro){no(ro)}},yo=oo=>oo.done?co(oo.value):Promise.resolve(oo.value).then(mo,bo);yo((to=to.apply(fo,po)).next())});var ao=()=>eo(void 0,null,function*(){return"foo"});var so=()=>eo(void 0,null,function*(){return"bar"});Promise.all([ao(),so()]).then(([fo,po])=>{console.warn(fo,po)});})();
",
]
`;

exports[`index esbuild-plugin-reserve-identifiers basic 1`] = `
[
  ""use strict";(()=>{var oo=()=>{console.log("a")},to=()=>{console.log("b")},eo=()=>{console.log("c")},yo=()=>{console.log("d")},fo=()=>{console.log("e")},po=()=>{console.log("f")},co=()=>{console.log("g")},lo=()=>{console.log("h")},no=()=>{console.log("i")},so=()=>{console.log("j")},go=()=>{console.log("k")},_o=()=>{console.log("l")},wo=()=>{console.log("m")},xo=()=>{console.log("n")},ao=()=>{console.log("o")},bo=()=>{console.log("p")},ho=()=>{console.log("q")},io=()=>{console.log("r")},jo=()=>{console.log("s")},ko=()=>{console.log("t")},mo=()=>{console.log("u")},qo=()=>{console.log("v")},ro=()=>{console.log("w")},uo=()=>{console.log("x")},vo=()=>{console.log("y")},zo=()=>{console.log("z")};oo();to();eo();yo();fo();po();co();lo();no();so();go();_o();wo();xo();ao();bo();ho();io();jo();ko();mo();qo();ro();uo();vo();zo();})();
",
]
`;

exports[`index esbuild-plugin-reserve-identifiers react 1`] = `
[
  ""use strict";(()=>{var io=Object.create;var ro=Object.defineProperty;var mo=Object.getOwnPropertyDescriptor;var so=Object.getOwnPropertyNames;var vo=Object.getPrototypeOf,ao=Object.prototype.hasOwnProperty;var no=(oo=>typeof require!="undefined"?require:typeof Proxy!="undefined"?new Proxy(oo,{get:(eo,to)=>(typeof require!="undefined"?require:eo)[to]}):oo)(function(oo){if(typeof require!="undefined")return require.apply(this,arguments);throw Error('Dynamic require of "'+oo+'" is not supported')});var uo=(oo,eo,to,fo)=>{if(eo&&typeof eo=="object"||typeof eo=="function")for(let po of so(eo))!ao.call(oo,po)&&po!==to&&ro(oo,po,{get:()=>eo[po],enumerable:!(fo=mo(eo,po))||fo.enumerable});return oo};var bo=(oo,eo,to)=>(to=oo!=null?io(vo(oo)):{},uo(eo||!oo||!oo.__esModule?ro(to,"default",{value:oo,enumerable:!0}):to,oo));var yo=bo(no("react")),lo=no("react-dom/client");var wo=()=>(console.warn("Hello World"),yo.default.createElement("div",null,"Hello World")),co=document.createElement("div");document.body.appendChild(co);var Ho=(0,lo.createRoot)(co);Ho.render(yo.default.createElement(wo,null));})();
",
  "body{color:#0074d9}
",
]
`;

exports[`index esbuild-plugin-reserve-identifiers react-jsx 1`] = `
[
  ""use strict";(()=>{var io=Object.create;var ro=Object.defineProperty;var mo=Object.getOwnPropertyDescriptor;var so=Object.getOwnPropertyNames;var vo=Object.getPrototypeOf,ao=Object.prototype.hasOwnProperty;var no=(oo=>typeof require!="undefined"?require:typeof Proxy!="undefined"?new Proxy(oo,{get:(eo,to)=>(typeof require!="undefined"?require:eo)[to]}):oo)(function(oo){if(typeof require!="undefined")return require.apply(this,arguments);throw Error('Dynamic require of "'+oo+'" is not supported')});var uo=(oo,eo,to,fo)=>{if(eo&&typeof eo=="object"||typeof eo=="function")for(let po of so(eo))!ao.call(oo,po)&&po!==to&&ro(oo,po,{get:()=>eo[po],enumerable:!(fo=mo(eo,po))||fo.enumerable});return oo};var bo=(oo,eo,to)=>(to=oo!=null?io(vo(oo)):{},uo(eo||!oo||!oo.__esModule?ro(to,"default",{value:oo,enumerable:!0}):to,oo));var yo=bo(no("react")),lo=no("react-dom/client");var wo=()=>(console.warn("Hello World"),yo.default.createElement("div",null,"Hello World")),co=document.createElement("div");document.body.appendChild(co);var Ho=(0,lo.createRoot)(co);Ho.render(yo.default.createElement(wo,null));})();
",
  "body{color:#0074d9}
",
]
`;
